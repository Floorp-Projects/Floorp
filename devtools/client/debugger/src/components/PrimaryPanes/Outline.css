/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.outline {
  overflow-y: hidden;
}

.outline > div {
  width: 100%;
  position: relative;
}

.outline-pane-info {
  padding: 0.5em;
  width: 100%;
  font-style: italic;
  text-align: center;
  user-select: none;
  font-size: 12px;
  overflow: hidden;
}

.outline-list {
  margin: 0;
  padding: 4px 0;
  position: absolute;
  top: 25px;
  bottom: 25px;
  left: 0;
  right: 0;
  list-style-type: none;
  overflow: auto;
}

.outline-list__class-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.outline-list__class-list > .outline-list__element {
  padding-inline-start: 2rem;
}

.outline-list__class-list .function-signature .function-name {
  color: var(--theme-highlight-green);
}

.outline-list .function-signature .paren {
  color: inherit;
}

.outline-list__class h2 {
  font-weight: normal;
  font-size: 1em;
  padding: 3px 0;
  padding-inline-start: 10px;
  color: var(--blue-55);
  margin: 0;
}

.outline-list__class:not(:first-child) h2 {
  margin-top: 12px;
}

.outline-list h2:hover {
  background: var(--theme-toolbar-background-hover);
}

.theme-dark .outline-list h2 {
  color: var(--theme-highlight-blue);
}

.outline-list h2 .keyword {
  color: var(--theme-highlight-red);
}

.outline-list__class h2.focused {
  background: var(--theme-selection-background);
}

.outline-list__class h2.focused,
.outline-list__class h2.focused .keyword {
  color: var(--theme-selection-color);
}

.outline-list__element {
  padding: 3px 10px 3px 10px;
  cursor: default;
  white-space: nowrap;
}

.outline-list > .outline-list__element {
  padding-inline-start: 1rem;
}

.outline-list__element-icon {
  padding-inline-end: 0.4rem;
}

.outline-list__element:hover {
  background: var(--theme-toolbar-background-hover);
}

.outline-list__element.focused {
  background: var(--theme-selection-background);
}

.outline-list__element.focused .outline-list__element-icon,
.outline-list__element.focused .function-signature * {
  color: var(--theme-selection-color);
}

.outline-footer {
  display: flex;
  box-sizing: border-box;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: 25px;
  background: var(--theme-body-background);
  border-top: 1px solid var(--theme-splitter-color);
  opacity: 1;
  z-index: 1;
  user-select: none;
}

.outline-footer button {
  color: var(--theme-body-color);
}

.outline-footer button.active {
  background: var(--theme-selection-background);
  color: var(--theme-selection-color);
}
