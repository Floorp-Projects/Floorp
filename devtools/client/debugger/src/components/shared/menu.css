/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

menupopup {
  position: fixed;
  z-index: 10000;
  background: white;
  border: 1px solid #cccccc;
  padding: 5px 0;
  background: #f2f2f2;
  border-radius: 5px;
  color: #585858;
  box-shadow: 0 0 4px 0 rgba(190, 190, 190, 0.8);
  min-width: 130px;
}

menuitem {
  display: block;
  padding: 0 20px;
  line-height: 20px;
  font-weight: 500;
  font-size: 13px;
  user-select: none;
}

menuitem:hover {
  background: #3780fb;
  color: white;
}

menuitem[disabled="true"] {
  color: #cccccc;
}

menuitem[disabled="true"]:hover {
  background-color: transparent;
  cursor: default;
}

menuseparator {
  border-bottom: 1px solid #cacdd3;
  width: 100%;
  height: 5px;
  display: block;
  margin-bottom: 5px;
}

#contextmenu-mask.show {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 999;
}
