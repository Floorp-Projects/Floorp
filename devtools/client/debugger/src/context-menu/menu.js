/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

const Menu = require("devtools/client/framework/menu");
const MenuItem = require("devtools/client/framework/menu-item");

export function showMenu(evt, items) {
  if (items.length === 0) {
    return;
  }

  const menu = new Menu();
  items
    .filter(item => item.visible === undefined || item.visible === true)
    .forEach(item => {
      const menuItem = new MenuItem(item);
      menuItem.submenu = createSubMenu(item.submenu);
      menu.append(menuItem);
    });

  menu.popup(evt.screenX, evt.screenY, window.parent.document);
}

function createSubMenu(subItems) {
  if (subItems) {
    const subMenu = new Menu();
    subItems.forEach(subItem => {
      subMenu.append(new MenuItem(subItem));
    });
    return subMenu;
  }
  return null;
}

export function buildMenu(items) {
  return items
    .map(itm => {
      const hide = typeof itm.hidden === "function" ? itm.hidden() : itm.hidden;
      return hide ? null : itm.item;
    })
    .filter(itm => itm !== null);
}
