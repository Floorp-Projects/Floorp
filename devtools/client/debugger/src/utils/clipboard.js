/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

/**
 * Clipboard function taken from
 * https://searchfox.org/mozilla-central/source/devtools/shared/platform/clipboard.js
 */

export function copyToTheClipboard(string) {
  const doCopy = function(e) {
    e.clipboardData.setData("text/plain", string);
    e.preventDefault();
  };

  document.addEventListener("copy", doCopy);
  document.execCommand("copy", false, null);
  document.removeEventListener("copy", doCopy);
}
