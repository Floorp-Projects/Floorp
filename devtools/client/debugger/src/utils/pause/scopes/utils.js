/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

export function getFramePopVariables(why, path) {
  const vars = [];

  if (why && why.frameFinished) {
    const { frameFinished } = why;

    // Always display a `throw` property if present, even if it is falsy.
    if (Object.prototype.hasOwnProperty.call(frameFinished, "throw")) {
      vars.push({
        name: "<exception>",
        path: `${path}/<exception>`,
        contents: { value: frameFinished.throw },
      });
    }

    if (Object.prototype.hasOwnProperty.call(frameFinished, "return")) {
      const returned = frameFinished.return;

      // Do not display undefined. Do display falsy values like 0 and false. The
      // protocol grip for undefined is a JSON object: { type: "undefined" }.
      if (typeof returned !== "object" || returned.type !== "undefined") {
        vars.push({
          name: "<return>",
          path: `${path}/<return>`,
          contents: { value: returned },
        });
      }
    }
  }

  return vars;
}

export function getThisVariable(this_, path) {
  if (!this_) {
    return null;
  }

  return {
    name: "<this>",
    path: `${path}/<this>`,
    contents: { value: this_ },
  };
}

// Get a string path for an scope item which can be used in different pauses for
// a thread.
export function getScopeItemPath(item) {
  // Calling toString() on item.path allows symbols to be handled.
  return item.path.toString();
}
