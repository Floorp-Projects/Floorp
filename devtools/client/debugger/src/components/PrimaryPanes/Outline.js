/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import { showMenu } from "../../context-menu/menu";
import { connect } from "../../utils/connect";
import { score as fuzzaldrinScore } from "fuzzaldrin-plus";
const classnames = require("classnames");

import { containsPosition, positionAfter } from "../../utils/ast";
import { copyToTheClipboard } from "../../utils/clipboard";
import { findFunctionText } from "../../utils/function";

import actions from "../../actions";
import {
  getSelectedSourceWithContent,
  getSymbols,
  getCursorPosition,
  getContext,
} from "../../selectors";

import OutlineFilter from "./OutlineFilter";
import "./Outline.css";
import PreviewFunction from "../shared/PreviewFunction";
import { uniq, sortBy } from "lodash";

// Set higher to make the fuzzaldrin filter more specific
const FUZZALDRIN_FILTER_THRESHOLD = 15000;

/**
 * Check whether the name argument matches the fuzzy filter argument
 */
const filterOutlineItem = (name, filter) => {
  if (!filter) {
    return true;
  }

  if (filter.length === 1) {
    // when filter is a single char just check if it starts with the char
    return filter.toLowerCase() === name.toLowerCase()[0];
  }
  return fuzzaldrinScore(name, filter) > FUZZALDRIN_FILTER_THRESHOLD;
};

// Checks if an element is visible inside its parent element
function isVisible(element, parent) {
  const parentRect = parent.getBoundingClientRect();
  const elementRect = element.getBoundingClientRect();

  const parentTop = parentRect.top;
  const parentBottom = parentRect.bottom;
  const elTop = elementRect.top;
  const elBottom = elementRect.bottom;

  return parentTop < elTop && parentBottom > elBottom;
}

export class Outline extends Component {
  constructor(props) {
    super(props);
    this.focusedElRef = null;
    this.state = { filter: "", focusedItem: null };
  }

  componentDidUpdate(prevProps) {
    const { cursorPosition, symbols } = this.props;
    if (
      cursorPosition &&
      symbols &&
      cursorPosition !== prevProps.cursorPosition
    ) {
      this.setFocus(cursorPosition);
    }

    if (
      this.focusedElRef &&
      !isVisible(this.focusedElRef, this.refs.outlineList)
    ) {
      this.focusedElRef.scrollIntoView({ block: "center" });
    }
  }

  setFocus(cursorPosition) {
    const { symbols } = this.props;
    let classes = [];
    let functions = [];

    if (symbols && !symbols.loading) {
      ({ classes, functions } = symbols);
    }

    // Find items that enclose the selected location
    const enclosedItems = [...classes, ...functions].filter(
      ({ name, location }) =>
        name != "anonymous" && containsPosition(location, cursorPosition)
    );

    if (enclosedItems.length == 0) {
      return this.setState({ focusedItem: null });
    }

    // Find the closest item to the selected location to focus
    const closestItem = enclosedItems.reduce((item, closest) =>
      positionAfter(item.location, closest.location) ? item : closest
    );

    this.setState({ focusedItem: closestItem });
  }

  selectItem(selectedItem) {
    const { cx, selectedSource, selectLocation } = this.props;
    if (!selectedSource || !selectedItem) {
      return;
    }

    selectLocation(cx, {
      sourceId: selectedSource.id,
      line: selectedItem.location.start.line,
      column: selectedItem.location.start.column,
    });

    this.setState({ focusedItem: selectedItem });
  }

  onContextMenu(event, func) {
    event.stopPropagation();
    event.preventDefault();

    const { selectedSource, flashLineRange, getFunctionText } = this.props;

    if (!selectedSource) {
      return;
    }

    const sourceLine = func.location.start.line;
    const functionText = getFunctionText(sourceLine);

    const copyFunctionItem = {
      id: "node-menu-copy-function",
      label: L10N.getStr("copyFunction.label"),
      accesskey: L10N.getStr("copyFunction.accesskey"),
      disabled: !functionText,
      click: () => {
        flashLineRange({
          start: sourceLine,
          end: func.location.end.line,
          sourceId: selectedSource.id,
        });
        return copyToTheClipboard(functionText);
      },
    };
    const menuOptions = [copyFunctionItem];
    showMenu(event, menuOptions);
  }

  updateFilter = filter => {
    this.setState({ filter: filter.trim() });
  };

  renderPlaceholder() {
    const placeholderMessage = this.props.selectedSource
      ? L10N.getStr("outline.noFunctions")
      : L10N.getStr("outline.noFileSelected");

    return <div className="outline-pane-info">{placeholderMessage}</div>;
  }

  renderLoading() {
    return (
      <div className="outline-pane-info">{L10N.getStr("loadingText")}</div>
    );
  }

  renderFunction(func) {
    const { focusedItem } = this.state;
    const { name, location, parameterNames } = func;
    const isFocused = focusedItem === func;

    return (
      <li
        key={`${name}:${location.start.line}:${location.start.column}`}
        className={classnames("outline-list__element", { focused: isFocused })}
        ref={el => {
          if (isFocused) {
            this.focusedElRef = el;
          }
        }}
        onClick={() => this.selectItem(func)}
        onContextMenu={e => this.onContextMenu(e, func)}
      >
        <span className="outline-list__element-icon">Î»</span>
        <PreviewFunction func={{ name, parameterNames }} />
      </li>
    );
  }

  renderClassHeader(klass) {
    return (
      <div>
        <span className="keyword">class</span> {klass}
      </div>
    );
  }

  renderClassFunctions(klass, functions) {
    const { symbols } = this.props;

    if (!symbols || symbols.loading || klass == null || functions.length == 0) {
      return null;
    }

    const { focusedItem } = this.state;
    const classFunc = functions.find(func => func.name === klass);
    const classFunctions = functions.filter(func => func.klass === klass);
    const classInfo = symbols.classes.find(c => c.name === klass);

    const item = classFunc || classInfo;
    const isFocused = focusedItem === item;

    return (
      <li
        className="outline-list__class"
        ref={el => {
          if (isFocused) {
            this.focusedElRef = el;
          }
        }}
        key={klass}
      >
        <h2
          className={classnames("", { focused: isFocused })}
          onClick={() => this.selectItem(item)}
        >
          {classFunc
            ? this.renderFunction(classFunc)
            : this.renderClassHeader(klass)}
        </h2>
        <ul className="outline-list__class-list">
          {classFunctions.map(func => this.renderFunction(func))}
        </ul>
      </li>
    );
  }

  renderFunctions(functions) {
    const { filter } = this.state;
    let classes = uniq(functions.map(({ klass }) => klass));
    let namedFunctions = functions.filter(
      ({ name, klass }) =>
        filterOutlineItem(name, filter) && !klass && !classes.includes(name)
    );

    let classFunctions = functions.filter(
      ({ name, klass }) => filterOutlineItem(name, filter) && !!klass
    );

    if (this.props.alphabetizeOutline) {
      namedFunctions = sortBy(namedFunctions, "name");
      classes = classes.sort();
      classFunctions = sortBy(classFunctions, "name");
    }

    return (
      <ul
        ref="outlineList"
        className="outline-list devtools-monospace"
        dir="ltr"
      >
        {namedFunctions.map(func => this.renderFunction(func))}
        {classes.map(klass => this.renderClassFunctions(klass, classFunctions))}
      </ul>
    );
  }

  renderFooter() {
    return (
      <div className="outline-footer">
        <button
          onClick={this.props.onAlphabetizeClick}
          className={this.props.alphabetizeOutline ? "active" : ""}
        >
          {L10N.getStr("outline.sortLabel")}
        </button>
      </div>
    );
  }

  render() {
    const { symbols, selectedSource } = this.props;
    const { filter } = this.state;

    if (!selectedSource) {
      return this.renderPlaceholder();
    }

    if (!symbols || symbols.loading) {
      return this.renderLoading();
    }

    const symbolsToDisplay = symbols.functions.filter(
      ({ name }) => name != "anonymous"
    );

    if (symbolsToDisplay.length === 0) {
      return this.renderPlaceholder();
    }

    return (
      <div className="outline">
        <div>
          <OutlineFilter filter={filter} updateFilter={this.updateFilter} />
          {this.renderFunctions(symbolsToDisplay)}
          {this.renderFooter()}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const selectedSource = getSelectedSourceWithContent(state);
  const symbols = selectedSource ? getSymbols(state, selectedSource) : null;

  return {
    cx: getContext(state),
    symbols,
    selectedSource: selectedSource,
    cursorPosition: getCursorPosition(state),
    getFunctionText: line => {
      if (selectedSource) {
        return findFunctionText(line, selectedSource, symbols);
      }

      return null;
    },
  };
};

export default connect(mapStateToProps, {
  selectLocation: actions.selectLocation,
  flashLineRange: actions.flashLineRange,
})(Outline);
