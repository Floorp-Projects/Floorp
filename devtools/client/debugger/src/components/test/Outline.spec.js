/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import Outline from "../../components/PrimaryPanes/Outline";
import { makeSymbolDeclaration } from "../../utils/test-head";
import { mockcx } from "../../utils/test-mockup";
import { showMenu } from "../../context-menu/menu";
import { copyToTheClipboard } from "../../utils/clipboard";

jest.mock("../../context-menu/menu", () => ({ showMenu: jest.fn() }));
jest.mock("../../utils/clipboard", () => ({ copyToTheClipboard: jest.fn() }));

const sourceId = "id";
const mockFunctionText = "mock function text";

function generateDefaults(overrides) {
  return {
    cx: mockcx,
    selectLocation: jest.fn(),
    selectedSource: { id: sourceId },
    getFunctionText: jest.fn().mockReturnValue(mockFunctionText),
    flashLineRange: jest.fn(),
    isHidden: false,
    symbols: {},
    selectedLocation: { sourceId },
    onAlphabetizeClick: jest.fn(),
    ...overrides,
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const component = shallow(<Outline.WrappedComponent {...props} />);
  const instance = component.instance();
  return { component, props, instance };
}

describe("Outline", () => {
  afterEach(() => {
    copyToTheClipboard.mockClear();
    showMenu.mockClear();
  });

  it("renders a list of functions when properties change", async () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("my_example_function1", 21),
        makeSymbolDeclaration("my_example_function2", 22),
      ],
    };

    const { component } = render({ symbols });
    expect(component).toMatchSnapshot();
  });

  it("selects a line of code in the current file on click", async () => {
    const startLine = 12;
    const symbols = {
      functions: [makeSymbolDeclaration("my_example_function", startLine)],
    };

    const { component, props } = render({ symbols });

    const { selectLocation } = props;
    const listItem = component.find("li").first();
    listItem.simulate("click");
    expect(selectLocation).toHaveBeenCalledWith(mockcx, {
      line: startLine,
      sourceId,
    });
  });

  describe("renders outline", () => {
    describe("renders loading", () => {
      it("if symbols is not defined", () => {
        const { component } = render({
          symbols: null,
        });
        expect(component).toMatchSnapshot();
      });

      it("if symbols are loading", () => {
        const { component } = render({
          symbols: {
            loading: true,
          },
        });
        expect(component).toMatchSnapshot();
      });
    });

    it("renders ignore anonymous functions", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("my_example_function1", 21),
          makeSymbolDeclaration("anonymous", 25),
        ],
      };

      const { component } = render({ symbols });
      expect(component).toMatchSnapshot();
    });
    describe("renders placeholder", () => {
      it("`No File Selected` if selectedSource is not defined", async () => {
        const { component } = render({
          selectedSource: null,
        });
        expect(component).toMatchSnapshot();
      });

      it("`No functions` if all func are anonymous", async () => {
        const symbols = {
          functions: [
            makeSymbolDeclaration("anonymous", 25),
            makeSymbolDeclaration("anonymous", 30),
          ],
        };

        const { component } = render({ symbols });
        expect(component).toMatchSnapshot();
      });

      it("`No functions` if symbols has no func", async () => {
        const symbols = {
          functions: [],
        };
        const { component } = render({ symbols });
        expect(component).toMatchSnapshot();
      });
    });

    it("sorts functions alphabetically by function name", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("c_function", 25),
          makeSymbolDeclaration("x_function", 30),
          makeSymbolDeclaration("a_function", 70),
        ],
      };

      const { component } = render({
        symbols,
        alphabetizeOutline: true,
      });
      expect(component).toMatchSnapshot();
    });

    it("calls onAlphabetizeClick when sort button is clicked", async () => {
      const symbols = {
        functions: [makeSymbolDeclaration("example_function", 25)],
      };

      const { component, props } = render({ symbols });

      await component
        .find(".outline-footer")
        .find("button")
        .simulate("click", {});

      expect(props.onAlphabetizeClick).toHaveBeenCalled();
    });

    it("renders functions by function class", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("x_function", 25, 26, "x_klass"),
          makeSymbolDeclaration("a2_function", 30, 31, "a_klass"),
          makeSymbolDeclaration("a1_function", 70, 71, "a_klass"),
        ],
        classes: [
          makeSymbolDeclaration("x_klass", 24, 27),
          makeSymbolDeclaration("a_klass", 29, 72),
        ],
      };

      const { component } = render({ symbols });
      expect(component).toMatchSnapshot();
    });

    it("renders functions by function class, alphabetically", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("x_function", 25, 26, "x_klass"),
          makeSymbolDeclaration("a2_function", 30, 31, "a_klass"),
          makeSymbolDeclaration("a1_function", 70, 71, "a_klass"),
        ],
        classes: [
          makeSymbolDeclaration("x_klass", 24, 27),
          makeSymbolDeclaration("a_klass", 29, 72),
        ],
      };

      const { component } = render({
        symbols,
        alphabetizeOutline: true,
      });
      expect(component).toMatchSnapshot();
    });

    it("selects class on click on class headline", async () => {
      const symbols = {
        functions: [makeSymbolDeclaration("x_function", 25, 26, "x_klass")],
        classes: [makeSymbolDeclaration("x_klass", 24, 27)],
      };

      const { component, props } = render({ symbols });

      await component.find("h2").simulate("click", {});

      expect(props.selectLocation).toHaveBeenCalledWith(mockcx, {
        line: 24,
        sourceId,
      });
    });

    it("does not select an item if selectedSource is not defined", async () => {
      const { instance, props } = render({ selectedSource: null });
      await instance.selectItem({});
      expect(props.selectLocation).not.toHaveBeenCalled();
    });
  });

  describe("onContextMenu of Outline", () => {
    it("is called onContextMenu for each item", async () => {
      const event = { event: "oncontextmenu" };
      const fn = makeSymbolDeclaration("exmple_function", 2);
      const symbols = {
        functions: [fn],
      };

      const { component, instance } = render({ symbols });
      instance.onContextMenu = jest.fn(() => {});
      await component
        .find(".outline-list__element")
        .simulate("contextmenu", event);

      expect(instance.onContextMenu).toHaveBeenCalledWith(event, fn);
    });

    it("does not show menu with no selected source", async () => {
      const mockEvent = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn(),
      };
      const { instance } = render({
        selectedSource: null,
      });
      await instance.onContextMenu(mockEvent, {});
      expect(mockEvent.preventDefault).toHaveBeenCalled();
      expect(mockEvent.stopPropagation).toHaveBeenCalled();
      expect(showMenu).not.toHaveBeenCalled();
    });

    it("shows menu to copy func, copies to clipboard on click", async () => {
      const startLine = 12;
      const endLine = 21;
      const func = makeSymbolDeclaration(
        "my_example_function",
        startLine,
        endLine
      );
      const symbols = {
        functions: [func],
      };
      const mockEvent = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn(),
      };
      const { instance, props } = render({ symbols });
      await instance.onContextMenu(mockEvent, func);

      expect(mockEvent.preventDefault).toHaveBeenCalled();
      expect(mockEvent.stopPropagation).toHaveBeenCalled();

      const expectedMenuOptions = [
        {
          accesskey: "F",
          click: expect.any(Function),
          disabled: false,
          id: "node-menu-copy-function",
          label: "Copy function",
        },
      ];
      expect(props.getFunctionText).toHaveBeenCalledWith(12);
      expect(showMenu).toHaveBeenCalledWith(mockEvent, expectedMenuOptions);

      showMenu.mock.calls[0][1][0].click();
      expect(copyToTheClipboard).toHaveBeenCalledWith(mockFunctionText);
      expect(props.flashLineRange).toHaveBeenCalledWith({
        end: endLine,
        sourceId,
        start: startLine,
      });
    });
  });
});
