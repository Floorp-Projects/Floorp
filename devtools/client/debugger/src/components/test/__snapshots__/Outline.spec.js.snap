// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outline renders a list of functions when properties change 1`] = `
<div
  className="outline"
>
  <div>
    <OutlineFilter
      filter=""
      updateFilter={[Function]}
    />
    <ul
      className="outline-list devtools-monospace"
      dir="ltr"
    >
      <li
        className="outline-list__element"
        key="my_example_function1:21:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "my_example_function1",
              "parameterNames": undefined,
            }
          }
        />
      </li>
      <li
        className="outline-list__element"
        key="my_example_function2:22:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "my_example_function2",
              "parameterNames": undefined,
            }
          }
        />
      </li>
    </ul>
    <div
      className="outline-footer"
    >
      <button
        className=""
        onClick={[MockFunction]}
      >
        Sort by name
      </button>
    </div>
  </div>
</div>
`;

exports[`Outline renders outline renders functions by function class 1`] = `
<div
  className="outline"
>
  <div>
    <OutlineFilter
      filter=""
      updateFilter={[Function]}
    />
    <ul
      className="outline-list devtools-monospace"
      dir="ltr"
    >
      <li
        className="outline-list__class"
        key="x_klass"
      >
        <h2
          className=""
          onClick={[Function]}
        >
          <div>
            <span
              className="keyword"
            >
              class
            </span>
             
            x_klass
          </div>
        </h2>
        <ul
          className="outline-list__class-list"
        >
          <li
            className="outline-list__element"
            key="x_function:25:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "x_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
        </ul>
      </li>
      <li
        className="outline-list__class"
        key="a_klass"
      >
        <h2
          className=""
          onClick={[Function]}
        >
          <div>
            <span
              className="keyword"
            >
              class
            </span>
             
            a_klass
          </div>
        </h2>
        <ul
          className="outline-list__class-list"
        >
          <li
            className="outline-list__element"
            key="a2_function:30:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "a2_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
          <li
            className="outline-list__element"
            key="a1_function:70:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "a1_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
        </ul>
      </li>
    </ul>
    <div
      className="outline-footer"
    >
      <button
        className=""
        onClick={[MockFunction]}
      >
        Sort by name
      </button>
    </div>
  </div>
</div>
`;

exports[`Outline renders outline renders functions by function class, alphabetically 1`] = `
<div
  className="outline"
>
  <div>
    <OutlineFilter
      filter=""
      updateFilter={[Function]}
    />
    <ul
      className="outline-list devtools-monospace"
      dir="ltr"
    >
      <li
        className="outline-list__class"
        key="a_klass"
      >
        <h2
          className=""
          onClick={[Function]}
        >
          <div>
            <span
              className="keyword"
            >
              class
            </span>
             
            a_klass
          </div>
        </h2>
        <ul
          className="outline-list__class-list"
        >
          <li
            className="outline-list__element"
            key="a1_function:70:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "a1_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
          <li
            className="outline-list__element"
            key="a2_function:30:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "a2_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
        </ul>
      </li>
      <li
        className="outline-list__class"
        key="x_klass"
      >
        <h2
          className=""
          onClick={[Function]}
        >
          <div>
            <span
              className="keyword"
            >
              class
            </span>
             
            x_klass
          </div>
        </h2>
        <ul
          className="outline-list__class-list"
        >
          <li
            className="outline-list__element"
            key="x_function:25:undefined"
            onClick={[Function]}
            onContextMenu={[Function]}
          >
            <span
              className="outline-list__element-icon"
            >
              λ
            </span>
            <PreviewFunction
              func={
                Object {
                  "name": "x_function",
                  "parameterNames": undefined,
                }
              }
            />
          </li>
        </ul>
      </li>
    </ul>
    <div
      className="outline-footer"
    >
      <button
        className="active"
        onClick={[MockFunction]}
      >
        Sort by name
      </button>
    </div>
  </div>
</div>
`;

exports[`Outline renders outline renders ignore anonymous functions 1`] = `
<div
  className="outline"
>
  <div>
    <OutlineFilter
      filter=""
      updateFilter={[Function]}
    />
    <ul
      className="outline-list devtools-monospace"
      dir="ltr"
    >
      <li
        className="outline-list__element"
        key="my_example_function1:21:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "my_example_function1",
              "parameterNames": undefined,
            }
          }
        />
      </li>
    </ul>
    <div
      className="outline-footer"
    >
      <button
        className=""
        onClick={[MockFunction]}
      >
        Sort by name
      </button>
    </div>
  </div>
</div>
`;

exports[`Outline renders outline renders loading if symbols are loading 1`] = `
<div
  className="outline-pane-info"
>
  Loading…
</div>
`;

exports[`Outline renders outline renders loading if symbols is not defined 1`] = `
<div
  className="outline-pane-info"
>
  Loading…
</div>
`;

exports[`Outline renders outline renders placeholder \`No File Selected\` if selectedSource is not defined 1`] = `
<div
  className="outline-pane-info"
>
  No file selected
</div>
`;

exports[`Outline renders outline renders placeholder \`No functions\` if all func are anonymous 1`] = `
<div
  className="outline-pane-info"
>
  No functions
</div>
`;

exports[`Outline renders outline renders placeholder \`No functions\` if symbols has no func 1`] = `
<div
  className="outline-pane-info"
>
  No functions
</div>
`;

exports[`Outline renders outline sorts functions alphabetically by function name 1`] = `
<div
  className="outline"
>
  <div>
    <OutlineFilter
      filter=""
      updateFilter={[Function]}
    />
    <ul
      className="outline-list devtools-monospace"
      dir="ltr"
    >
      <li
        className="outline-list__element"
        key="a_function:70:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "a_function",
              "parameterNames": undefined,
            }
          }
        />
      </li>
      <li
        className="outline-list__element"
        key="c_function:25:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "c_function",
              "parameterNames": undefined,
            }
          }
        />
      </li>
      <li
        className="outline-list__element"
        key="x_function:30:undefined"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        <span
          className="outline-list__element-icon"
        >
          λ
        </span>
        <PreviewFunction
          func={
            Object {
              "name": "x_function",
              "parameterNames": undefined,
            }
          }
        />
      </li>
    </ul>
    <div
      className="outline-footer"
    >
      <button
        className="active"
        onClick={[MockFunction]}
      >
        Sort by name
      </button>
    </div>
  </div>
</div>
`;
